addon-chrome/ChromePlacesProvider.js

```diff
@@ -0,0 +1,379 @@
+   * @param {Number} options.maxResults - Maximum number of history items to return. Default to 20
+   *
+   * @returns {Object} Promise that resolves with histories
+   */
+  static getHistory(options) {
+    const today = new Date().getTime();
+    const aYear = 365 * 24 * 60 * 60 * 1000;
+    const aYearAgo = today - aYear;
+    const maxResults = 20;
+    const defaultOption = {
+      text: "",
+      startTime: aYearAgo,
+      endTime: today,
+      maxResults
+    };
+    const searchOptions = options ? Object.assign(defaultOption, options) : defaultOption;
+    const startTime = searchOptions.startTime;
+    const endTime = searchOptions.endTime;
+    return new Promise((resolve) => {
+      chrome.history.search(searchOptions, (results) => {
+        if (startTime && endTime) {
+          // api uses start and end time as OR instead of AND
+          // so filter it for now or we could just use this ???
+          results = results.filter((result) => result.lastVisitTime > startTime && result.lastVisitTime < endTime);
+        }
+        this._getBookmark()
+          .then((bookmarks) => {
+            const histories = results.map((result) => this._transformHistory(result, bookmarks));
+            this._filterBlockedUrls(histories).then(resolve);
+          });
+      });
+    });
+  }
+
+  /**
+   * Rank history items (excluding google searches and localhost sites) by frecency score, then lastVisitTime, then url
+   *
+   * @param {Array} histories - List of history items
+   *
+   * @returns {Object} Promise that resolves with list of history items ranked by their frencency, lastVisitTime, url
+   */
+  static _getTopFrecentSites(histories) {
+    return new Promise((resolve) => {
+      const rows = histories
+        .filter((hist) => !/(google|localhost)/.test(hist.url))
+        .map((hist) => {
+          const frencency = this._calculateFrecency(hist);
+          return Object.assign(hist, {frencency});
+        })
+        .sort((a, b) => {
+          if (a.frencency > b.frencency) {
+            return -1;
+          }
+          if (a.frencency < b.frencency) {
+            return 1;
+          }
+          if (a.lastVisitTime > b.lastVisitTime) {
+            return -1;
+          }
+          if (a.url > b.url) {
+            return -1;
+          }
+          return 0;
+        });
+
+      resolve(rows);
+    });
+  }
+
+  /**
+   * Calculate frecency scores a history items base on
+   * https://dxr.mozilla.org/mozilla-central/source/mobile/android/base/java/org/mozilla/gecko/db/BrowserContract.java#124
+   * numVisits * max(1, 100 * 225 / (age*age + 225))
+   *
+   * @param {Object} hist History item to calculate frecency score for
+   *
+   * @returns {Number} The calculated frecency for the history item
+   */
+  static _calculateFrecency(hist) {
+    const microsecondsPerDay = 86400000000;
+    const age = (new Date().getTime() - hist.lastVisitDate) / microsecondsPerDay;
```

Comment:
getTime() is milliseconds; not sure about `hist.lastVisitDate`, but probably also milliseconds. Don't mix units.